### PR TITLE
[SYCL][OpenCL] Enable graph extension on OpenCL backend

### PR DESCRIPTION
**Pull request by martygrant**
intel-llvm CI run for adding Command Buffers to the OpenCL Adapter in Unified Runtime - https://github.com/oneapi-src/unified-runtime/pull/966

Also completes follow-on work identified in #11599 to add an OpenCL section to the SYCL-Graphs docs and update the e2e Graph tests. Updating the tests has since been completed in a separate PR - https://github.com/intel/llvm/pull/11877

Depends on https://github.com/intel/llvm/pull/11820 merging first.

### CODE REVIEW

File: sycl/test-e2e/Graph/Explicit/assume_buffer_outlives_graph_property.cpp
Comment (EwanC) [inline sycl/test-e2e/Graph/Explicit/assume_buffer_outlives_graph_property.cpp:2]: Reckon we could drop the gpu from this and shorten both lines to a single `REQUIRES: level_zero || opencl`

File: sycl/test-e2e/Graph/graph_exception_global_device_extension.cpp
Comment (EwanC) [inline sycl/test-e2e/Graph/graph_exception_global_device_extension.cpp:150]: Can we call this `Queue` and pass it to `test()` as a parameter, rather than creating a new queue in `test()`

File: sycl/doc/design/CommandGraph.md
Comment (EwanC) [inline sycl/doc/design/CommandGraph.md:209]: Can we keep lines to 80 characters long

Comment (EwanC) [inline sycl/doc/design/CommandGraph.md:209]: Can we link to the `cl_khr_command_buffer` spec in this intro

File: sycl/doc/design/CommandGraph.md
Comment (EwanC) [inline sycl/doc/design/CommandGraph.md:234]: * "Append" -> "append"

File: sycl/doc/design/CommandGraph.md
Comment (EwanC) [inline sycl/doc/design/CommandGraph.md:236]: The line numbers used in this cl_ext.h URL won't be accurate as the file changes over time, just link to the file without a particular line

Comment (EwanC) [reply sycl/doc/design/CommandGraph.md:236]: "Future peice" -> "future piece"

File: sycl/doc/design/CommandGraph.md
Comment (EwanC) [inline sycl/doc/design/CommandGraph.md:235]: Can this section document what the gaps in SYCL support are. e.g Can't use memcpy operations in a graph node. This should reflect the features used in E2E test we haven't enabled for OpenCL.

File: sycl/doc/design/CommandGraph.md
Comment (EwanC) [inline sycl/doc/design/CommandGraph.md:243]: Can you link the command-buffer emulation layer as well https://github.com/bashbaug/SimpleOpenCLSamples/tree/main/layers/10_cmdbufemu

File: sycl/test-e2e/Graph/Explicit/usm_fill.cpp
Comment (EwanC) [inline sycl/test-e2e/Graph/Explicit/usm_fill.cpp:1]: Should we add level zero & opencl back here

File: sycl/doc/design/CommandGraph.md
Comment (reble) [inline sycl/doc/design/CommandGraph.md:234]:
```suggestion
Many of the OpenCL functions take a `cl_command_queue` parameter which is not present in most of the UR functions. Instead, when a new command buffer is created in `urCommandBufferCreateExp` we also create and maintain a new internal `ur_queue_handle_t` with a reference stored inside of the `ur_exp_command_buffer_handle_t_` struct. This internal queue is then used with the various append functions. The internal queue is retained and released whenever the owning command buffer is retained or released.
```

File: sycl/doc/design/CommandGraph.md
Comment (reble) [inline sycl/doc/design/CommandGraph.md:236]:
```suggestion
With command buffers being an OpenCL extension, each function is accessed by loading a function pointer to its implementation. These are defined in a common header file in the UR OpenCL adapter. The symbols for the functions are however defined in [OpenCL-Headers](https://github.com/KhronosGroup/OpenCL-Headers/blob/main/CL/cl_ext.h#L102-L110) but it is not known at this time what version of the headers will be used in the UR GitHub CI configuration, so loading the function pointers will be used until this can be verified. A future piece of work would be replacing the custom defined symbols with the ones from OpenCL-Headers.
```

File: sycl/test-e2e/Graph/RecordReplay/sub_graph.cpp
Comment (reble) [inline sycl/test-e2e/Graph/RecordReplay/sub_graph.cpp:11]: This change looks unrelated to the proposed patch!

File: sycl/doc/design/CommandGraph.md
Comment (EwanC) [inline sycl/doc/design/CommandGraph.md:232]: Can we include these OpenCL SVM commands as well in the table https://github.com/KhronosGroup/OpenCL-Docs/pull/915

File: sycl/test-e2e/Graph/Explicit/assume_buffer_outlives_graph_property.cpp
Comment (martygrant) [reply sycl/test-e2e/Graph/Explicit/assume_buffer_outlives_graph_property.cpp]: I'm quite unfamiliar with `lit` and was unsure how to use conditionals here. I was trying `opencl || level_zero, gpu` but that was failing so found the above change worked but removing `gpu` makes things simpler and have changed all the tests to that. Thanks!

File: sycl/test-e2e/Graph/graph_exception_global_device_extension.cpp
Comment (martygrant) [reply sycl/test-e2e/Graph/graph_exception_global_device_extension.cpp]: Updated thanks, made this consistent across all the tests.

File: sycl/doc/design/CommandGraph.md
Comment (martygrant) [reply sycl/doc/design/CommandGraph.md]: Updated thanks.

File: sycl/doc/design/CommandGraph.md
Comment (martygrant) [reply sycl/doc/design/CommandGraph.md]: Updated thanks.

File: sycl/doc/design/CommandGraph.md
Comment (martygrant) [reply sycl/doc/design/CommandGraph.md]: Updated thanks.

File: sycl/test-e2e/Graph/Explicit/usm_fill.cpp
Comment (martygrant) [reply sycl/test-e2e/Graph/Explicit/usm_fill.cpp]: Updated thanks. This was unintentional.

File: sycl/test-e2e/Graph/RecordReplay/sub_graph.cpp
Comment (martygrant) [reply sycl/test-e2e/Graph/RecordReplay/sub_graph.cpp:11]: I'm not quite sure why this is being highlighted as it's [unchanged from the current revision](https://github.com/intel/llvm/blob/sycl/sycl/test-e2e/Graph/RecordReplay/sub_graph.cpp).

File: sycl/doc/design/CommandGraph.md
Comment (martygrant) [reply sycl/doc/design/CommandGraph.md:236]: Added these, thanks.

**Comment by martygrant** [comment]
In a rebased force-push I have corrected the `int Pattern = 3.14f` lines in the `usm_fill` tests, as well as enabling a few more tests for OpenCL as a bug in the `urCommandBufferAppendKernelLaunchExp` was corrected and will be updated in the [PR for it in the UR repo](https://github.com/oneapi-src/unified-runtime/pull/966).

File: sycl/test-e2e/Graph/RecordReplay/sub_graph.cpp
Comment (Bensuo) [reply sycl/test-e2e/Graph/RecordReplay/sub_graph.cpp:11]: I think this might be that a newline at EOF was added, but Github doesn't directly show them. In the old diff it indicates this was missing with the red warning icon.

Comment (martygrant) [reply sycl/test-e2e/Graph/RecordReplay/sub_graph.cpp:11]: Ah right that makes sense, will just leave it as is then. Thanks Ben.

**Comment by martygrant** [comment]
Using this PR as the companion to the UR PR https://github.com/oneapi-src/unified-runtime/pull/966 instead of the separate https://github.com/intel/llvm/pull/11719 which I will now close.

File: sycl/doc/design/CommandGraph.md
Comment (EwanC) [reply sycl/doc/design/CommandGraph.md]: To elaborate on this point, because I think one of the most important parts of this PR is that we effectively communicate what's supported and what's not.

* Can you make this limitation information I've asked for a new sub-section in the document, because we'll probably need to link it to people.
* Could you also document how the gaps fail, because I'd like to make sure that exceptions or asserts are thrown (i'd guess in finalize()) when unsupported functionality is used. We want to to give a good user experience, rather than the segfaulting or the end result being subtly wrong.

**Comment by EwanC** [comment]
The Graph E2E tests added by this PR will also need updated as the other tests in this PR have been https://github.com/intel/llvm/pull/11556

File: sycl/test-e2e/Graph/device_query.cpp
Comment (EwanC) [inline sycl/test-e2e/Graph/device_query.cpp:1]: Wonder if we could remove this requires all together as well.

**Review by EwanC — COMMENTED** [review]
Minor comments after reading through the PR again, looks ready to go other than this though.

File: sycl/doc/design/CommandGraph.md
Comment (martygrant) [reply sycl/doc/design/CommandGraph.md]: Fixed, thanks.

File: sycl/doc/design/CommandGraph.md
Comment (martygrant) [reply sycl/doc/design/CommandGraph.md]: Fixed, thanks.

File: sycl/test-e2e/Graph/device_query.cpp
Comment (martygrant) [reply sycl/test-e2e/Graph/device_query.cpp]: Yeah this should have been removed with the other tests. Fixed now, cheers.

File: sycl/plugins/unified_runtime/CMakeLists.txt
Comment (kbenzie) [inline sycl/plugins/unified_runtime/CMakeLists.txt:65]:
```suggestion
  set(UNIFIED_RUNTIME_REPO "https://github.com/oneapi-src/unified-runtime.git")
  # commit 109ed46ee55f41317d35b2a9a20fa7a2029e9e64
  # Merge: 31b654f9 23005313
  # Author: Kenneth Benzie (Benie) <k.benzie@codeplay.com>
  # Date:   Wed Nov 22 16:04:52 2023 +0000
  #     Merge pull request #966 from martygrant/martin/openclCommandBuffers
  #     [OpenCL] Add Command Buffer extension to OpenCL adapter.
  set(UNIFIED_RUNTIME_TAG 109ed46ee55f41317d35b2a9a20fa7a2029e9e64)
```

**Review by kbenzie — COMMENTED** [review]
https://github.com/oneapi-src/unified-runtime/pull/966 has now been merged however this PR depends on changes in https://github.com/intel/llvm/pull/11820 so we should hold off updating the `UNIFIED_RUNTIME_REPO`/`UNIFIED_RUNTIME_TAG` values as suggested below for now.

File: sycl/plugins/unified_runtime/CMakeLists.txt
Comment (kbenzie) [reply sycl/plugins/unified_runtime/CMakeLists.txt:65]: https://github.com/intel/llvm/pull/11820 has merged now. I'll go ahead and make these changes and push them to this branch.

**Comment by kbenzie** [comment]
@intel/llvm-reviewers-runtime please review, this is intended for the next release.

**Review by aelovikov-intel — APPROVED** [review]

File: sycl/doc/design/CommandGraph.md
Comment (reble) [inline sycl/doc/design/CommandGraph.md:252]:
```suggestion
SYCL-Graph is only enabled for an OpenCL backend when the
```

File: sycl/doc/design/CommandGraph.md
Comment (reble) [inline sycl/doc/design/CommandGraph.md:267]:
```suggestion
SYCL backend cannot fully support the graphs API. Instead, there are
```

File: sycl/doc/design/CommandGraph.md
Comment (reble) [inline sycl/doc/design/CommandGraph.md:320]:
```suggestion
We are looking to address these gaps in the future so that SYCL-Graph can be
```

File: sycl/doc/design/CommandGraph.md
Comment (reble) [inline sycl/doc/design/CommandGraph.md:261]:
```suggestion
`device.get_info<info::device::graph_support>()` query for graph extension
```

File: sycl/doc/design/CommandGraph.md
Comment (reble) [inline sycl/doc/design/CommandGraph.md:259]:
```suggestion
to indicate that the graph extension should be enabled. This is information
```

File: sycl/doc/design/CommandGraph.md
Comment (reble) [inline sycl/doc/design/CommandGraph.md:348]: Can we find a more persistent link here?

Comment (kbenzie) [reply sycl/doc/design/CommandGraph.md:348]: Here's a permalink https://github.com/bashbaug/SimpleOpenCLSamples/tree/efeae73139ddf064fafce565cc39640af10d900f/layers/10_cmdbufemu

File: sycl/doc/design/CommandGraph.md
Comment (martygrant) [reply sycl/doc/design/CommandGraph.md]: Updated, thanks.

File: sycl/doc/design/CommandGraph.md
Comment (martygrant) [reply sycl/doc/design/CommandGraph.md]: Updated, thanks.

File: sycl/doc/design/CommandGraph.md
Comment (martygrant) [reply sycl/doc/design/CommandGraph.md]: Updated, thanks.

File: sycl/doc/design/CommandGraph.md
Comment (martygrant) [reply sycl/doc/design/CommandGraph.md]: Updated, thanks.

File: sycl/doc/design/CommandGraph.md
Comment (martygrant) [reply sycl/doc/design/CommandGraph.md]: Updated, thanks.

File: sycl/doc/design/CommandGraph.md
Comment (martygrant) [reply sycl/doc/design/CommandGraph.md]: Updated, thanks.

**Comment by kbenzie** [comment]
@intel/llvm-gatekeepers please merge.

**Comment by sergey-semenov** [comment]
@kbenzie This still needs approval from @intel/unified-runtime-reviewers

**Comment by kbenzie** [comment]
> @kbenzie This still needs approval from @intel/unified-runtime-reviewers

how foolish of me! approve now.

**Pull request merged**
